src/Language/Java/Inline/Magic.hs

```diff
@@ -0,0 +1,74 @@
+-- | Internal module defining some magic, kept separate from the rest, that
+-- depends on compiler internals.
+
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+module Language.Java.Inline.Magic
+  ( DotClass(className, classBytecode)
+  , mkDotClass
+  , isDotClassStaticKey
+  ) where
+
+import qualified Data.ByteString as BS
+import Foreign.Storable (sizeOf)
+import GHC.Exts (Int(..), Word(..), Word#)
+import GHC.Prim (indexWordArray#, sizeofByteArray#, unpackClosure#)
+import GHC.StaticPtr (StaticPtr, deRefStaticPtr)
+import qualified Language.Haskell.TH.Syntax as TH
+import Prelude hiding (words)
+
+-- | A wrapper for class bytecode. This datatype includes an unpacked
+-- fingerprint. This fingerprint is used by 'isDotClassStaticKey' to
+-- discriminate between static pointers to 'DotClass' values vs static pointers
+-- to other values just by inspecting whether the closure contains the magic
+-- values. Use 'mkDotClass' smart constructor to build.
+data DotClass = DotClass
+  { className :: String
+  , classBytecode :: BS.ByteString
+  , classMagicHigh :: Word#
+  , classMagicLow :: Word#
+  }
+
+mkDotClass
+  :: String -- ^ Class name
+  -> BS.ByteString -- ^ Class bytecode
+  -> DotClass
+mkDotClass className classBytecode =
+    let ![W# classMagicHigh, W# classMagicLow] = magic
+    in DotClass{..}
```

Comment:
I'm not sure I see what makes a `let-in` more convoluted than a `case-of`, but yes, that's the idea. Note that the `!` is forced by the fact that we're binding unboxed variables in the pattern.